### DISCUSSION TITLE
Computing Vector Electric Field from grad V

### DISCUSSION THREAD

**Discussion by jsevic** [idaholab/moose#25993] [Q&A General] [answered]
I think I should know this by now, I have voltage V as a state variable and I want to compute E = grad V as a vector so I can use in Paraview. I have this in an AuxKernel I get a compile error as follows...

 error: no viable conversion from returned value of type 'const libMesh::VectorValue<double>' to function return type 'libMesh::Real' (aka 'double')
  return  _grad_V[_qp];
          ^~~~~~~~~~~~
/Users/johnsevic/projects/xenon/src/auxkernels/ElectricFieldVector.C:47:10: error: no viable conversion from returned value of type 'const libMesh::VectorValue<double>' to function return type 'libMesh::Real' (aka 'double')
  return _grad_V[_qp];

In my .h file I declared as follows...

```
  const VariableGradient & _grad_V;
  const VariableValue & _D;
```

In my .c file I have the following...

```
InputParameters
ElectricFieldVector::validParams()
{
  InputParameters params = Kernel::validParams();
params.addRequiredCoupledVar("coupled_V", "grad V is E");
params.addRequiredCoupledVar("coupled_sigma", "The sigma variable");

  return params;
}

ElectricFieldVector::ElectricFieldVector(const InputParameters & parameters)
  : Kernel(parameters),
    _grad_V(coupledGradient("coupled_V")),
    _D(coupledValue("coupled_sigma"))
{
}

Real
ElectricFieldVector::computeQpResidual()
{
  return  _grad_V[_qp];
}

Real
ElectricFieldVector::computeQpJacobian()
{
  return _grad_V[_qp];
}
```

Thank you!

**Comment by cticenhour** [comment, marked as answer]
The error you are seeing is because the `computeQpResidual` method wants a `Real` number, but you are attempting to return a `RealVectorValue`. 

If you are only wanting E for visualization, I would compute this as a `VectorAuxKernel` instead of inheriting from `Kernel`. Then, you'd only need to define a `computeValue` method (removing the `Residual` and `Jacobian` methods you have defined). 

Did you mean in your original message that you wanted to compute E = - grad V? That's the normal form I see it in. Thus `computeValue` would simply be

```
RealVectorValue
ElectricFieldVector::computeValue()
{
   return -_grad_V[_qp];
}
```

**Reply by jsevic** [reply]
That works thanks!
